File: agents/backend/onyx/server/features/robot_maintenance_ai/core/plugin_system.py
"""
Plugin system for extending functionality.
"""

from typing import Dict, List, Optional, Any, Callable
from abc import ABC, abstractmethod
import logging

logger = logging.getLogger(__name__)


class Plugin(ABC):
    """Base class for plugins."""
    
    def __init__(self, name: str, version: str = "1.0.0"):
        self.name = name
        self.version = version
        self.enabled = True
    
    @abstractmethod
    def initialize(self, config: Dict[str, Any]) -> bool:
        """
        Initialize the plugin.
        
        Args:
            config: Plugin configuration
        
        Returns:
            True if initialization successful
        """
        pass
    
    @abstractmethod
    def execute(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Execute plugin logic.
        
        Args:
            data: Input data
        
        Returns:
            Output data
        """
        pass
    
    def cleanup(self):
        """Cleanup resources."""
        pass


class PluginManager:
    """Manages plugins for the system."""
    
    def __init__(self):
        self.plugins: Dict[str, Plugin] = {}
        self.hooks: Dict[str, List[Callable]] = {}
    
    def register_plugin(self, plugin: Plugin):
        """
        Register a plugin.
        
        Args:
            plugin: Plugin instance
        """
        self.plugins[plugin.name] = plugin
        logger.info(f"Plugin registered: {plugin.name} v{plugin.version}")
    
    def unregister_plugin(self, plugin_name: str):
        """
        Unregister a plugin.
        
        Args:
            plugin_name: Name of plugin to unregister
        """
        if plugin_name in self.plugins:
            plugin = self.plugins[plugin_name]
            plugin.cleanup()
            del self.plugins[plugin_name]
            logger.info(f"Plugin unregistered: {plugin_name}")
    
    def register_hook(self, hook_name: str, callback: Callable):
        """
        Register a hook callback.
        
        Args:
            hook_name: Name of the hook
            callback: Callback function
        """
        if hook_name not in self.hooks:
            self.hooks[hook_name] = []
        self.hooks[hook_name].append(callback)
        logger.info(f"Hook registered: {hook_name}")
    
    def execute_hook(self, hook_name: str, data: Any = None) -> List[Any]:
        """
        Execute all callbacks for a hook.
        
        Args:
            hook_name: Name of the hook
            data: Data to pass to callbacks
        
        Returns:
            List of results from callbacks
        """
        results = []
        if hook_name in self.hooks:
            for callback in self.hooks[hook_name]:
                try:
                    result = callback(data) if data is not None else callback()
                    results.append(result)
                except Exception as e:
                    logger.error(f"Error executing hook {hook_name}: {e}")
        return results
    
    def get_plugin(self, plugin_name: str) -> Optional[Plugin]:
        """
        Get a plugin by name.
        
        Args:
            plugin_name: Name of plugin
        
        Returns:
            Plugin instance or None
        """
        return self.plugins.get(plugin_name)
    
    def list_plugins(self) -> List[Dict[str, Any]]:
        """
        List all registered plugins.
        
        Returns:
            List of plugin information
        """
        return [
            {
                "name": plugin.name,
                "version": plugin.version,
                "enabled": plugin.enabled
            }
            for plugin in self.plugins.values()
        ]

File: character_clothing_changer_ai/static/js/core/plugin-system.js
/**
 * Plugin System
 * =============
 * Plugin system for extending application functionality
 */

const PluginSystem = {
    /**
     * Registered plugins
     */
    plugins: new Map(),
    
    /**
     * Plugin hooks
     */
    hooks: new Map(),
    
    /**
     * Register a plugin
     */
    register(name, plugin) {
        if (this.plugins.has(name)) {
            if (typeof Logger !== 'undefined') {
                Logger.warn(`Plugin ${name} is already registered`);
            }
            return false;
        }
        
        // Validate plugin
        if (!plugin.name || !plugin.version) {
            if (typeof Logger !== 'undefined') {
                Logger.error(`Plugin ${name} is missing required fields`);
            }
            return false;
        }
        
        // Initialize plugin
        if (plugin.init) {
            try {
                plugin.init();
            } catch (e) {
                if (typeof Logger !== 'undefined') {
                    Logger.error(`Failed to initialize plugin ${name}:`, e);
                }
                return false;
            }
        }
        
        this.plugins.set(name, plugin);
        
        // Register hooks
        if (plugin.hooks) {
            Object.keys(plugin.hooks).forEach(hookName => {
                this.registerHook(hookName, plugin.hooks[hookName]);
            });
        }
        
        if (typeof Logger !== 'undefined') {
            Logger.info(`Plugin registered: ${name} v${plugin.version}`);
        }
        
        if (typeof EventBus !== 'undefined') {
            EventBus.emit('plugin:registered', name, plugin);
        }
        
        return true;
    },
    
    /**
     * Unregister a plugin
     */
    unregister(name) {
        if (!this.plugins.has(name)) {
            return false;
        }
        
        const plugin = this.plugins.get(name);
        
        // Cleanup plugin
        if (plugin.cleanup) {
            try {
                plugin.cleanup();
            } catch (e) {
                if (typeof Logger !== 'undefined') {
                    Logger.error(`Failed to cleanup plugin ${name}:`, e);
                }
            }
        }
        
        // Remove hooks
        if (plugin.hooks) {
            Object.keys(plugin.hooks).forEach(hookName => {
                this.unregisterHook(hookName, plugin.hooks[hookName]);
            });
        }
        
        this.plugins.delete(name);
        
        if (typeof Logger !== 'undefined') {
            Logger.info(`Plugin unregistered: ${name}`);
        }
        
        if (typeof EventBus !== 'undefined') {
            EventBus.emit('plugin:unregistered', name);
        }
        
        return true;
    },
    
    /**
     * Register a hook
     */
    registerHook(hookName, callback) {
        if (!this.hooks.has(hookName)) {
            this.hooks.set(hookName, []);
        }
        
        this.hooks.get(hookName).push(callback);
    },
    
    /**
     * Unregister a hook
     */
    unregisterHook(hookName, callback) {
        if (!this.hooks.has(hookName)) {
            return;
        }
        
        const callbacks = this.hooks.get(hookName);
        const index = callbacks.indexOf(callback);
        if (index > -1) {
            callbacks.splice(index, 1);
        }
    },
    
    /**
     * Execute hooks
     */
    executeHook(hookName, ...args) {
        if (!this.hooks.has(hookName)) {
            return args[0]; // Return first argument if no hooks
        }
        
        let result = args[0];
        const callbacks = this.hooks.get(hookName);
        
        for (const callback of callbacks) {
            try {
                const callbackResult = callback(result, ...args.slice(1));
                if (callbackResult !== undefined) {
                    result = callbackResult;
                }
            } catch (e) {
                if (typeof Logger !== 'undefined') {
                    Logger.error(`Hook error in ${hookName}:`, e);
                }
            }
        }
        
        return result;
    },
    
    /**
     * Get plugin
     */
    getPlugin(name) {
        return this.plugins.get(name);
    },
    
    /**
     * Get all plugins
     */
    getAllPlugins() {
        return Array.from(this.plugins.values());
    },
    
    /**
     * Check if plugin is registered
     */
    hasPlugin(name) {
        return this.plugins.has(name);
    },
    
    /**
     * Get plugin count
     */
    getPluginCount() {
        return this.plugins.size;
    }
};

// Export for use
if (typeof module !== 'undefined' && module.exports) {
    module.exports = PluginSystem;
}

